crates/derive/src/combine_api.rs

```diff
@@ -54,6 +56,13 @@ impl StatsAccumulator for CombineStats {
 struct State {
     // Combiner which is doing the heavy lifting.
     combiner: doc::Combiner,
+    // Inferred shape of drained documents
+    shape: doc::inference::Shape,
+    // We need to keep track of the potentially-modified shape
```

Comment:
`doc::inference::Shape` => `doc::shape::Shape` or just `doc::Shape`.

i think we can drop `scratch_shape`, because you can instead check whether any `widen()` operation returned true to track whether it's changed?

Reply:
Yup, changed `shape` to `Option<Shape>`, and added a `shape_changed` flag on State.